lib/nimble_csv.ex

```diff
@@ -256,38 +259,51 @@ defmodule NimbleCSV do
       end
 
       defp init_parser(opts) do
-        state = if Keyword.has_key?(opts, :headers) do
-          IO.warn "the :headers option is deprecated, please use :skip_headers instead"
-          if Keyword.get(opts, :headers, true), do: :header, else: :line
-        else
-          if Keyword.get(opts, :skip_headers, true), do: :header, else: :line
-        end
+        initial_mode =
+          if Keyword.has_key?(opts, :headers) do
+            IO.warn("the :headers option is deprecated, please use :skip_headers instead")
+            if Keyword.get(opts, :headers, true), do: :header, else: :line
+          else
+            if Keyword.get(opts, :skip_headers, true), do: :header, else: :line
+          end
+
+        state = state(mode: initial_mode, line_number: 0)
 
         {state, :binary.compile_pattern(@separator), :binary.compile_pattern(@escape)}
       end
 
-      defp finalize_parser({:escape, _, _, _}) do
-        raise ParseError, "expected escape character #{@escape} but reached the end of file"
+      defp finalize_parser({:escape, _, _, state}) do
+        raise ParseError,
+          message: "expected escape character #{@escape} but reached the end of file",
+          line_number: state(state, :line_number)
       end
 
       defp finalize_parser(_) do
         :ok
       end
 
+      defp to_enum({state(mode: :line) = state, row}) do
+        {[row], state}
+      end
+
+      defp to_enum({state(mode: :header) = state, _}) do
+        {[], state(state, mode: :line)}
+      end
+
       defp to_enum(result) do
-        case result do
-          {:line, row} -> {[row], :line}
-          {:header, _} -> {[], :line}
-          {:escape, _, _, _} = escape -> {[], escape}
-        end
+        {[], result}
       end
 
       defp parse(line, {:escape, entry, row, state}, separator, escape) do
-        to_enum(escape(line, entry, row, state, separator, escape))
+        to_enum(escape(line, entry, row, inc_line_number(state), separator, escape))
```

Comment:
This is only true if the `@escape` is a newline, which is not always true. So I would recommend defining `inc_line_number` to increment only if @escape contains the newline. Otherwise, line_number should always be nil, and we wouldn't use it in error messages.

Can you please change it accordingly? Thank you!

Reply:
`line_number` counts lines in the input stream, so in the following example

```
name,last,year
john,doe,"1986
on"e,two,three
```

it will correctly report the error on line 3:

```
%NimbleCSV.ParseError{
  message: ~s(unexpected escape character " in "on\\"e,two,three\\n"),
  line_number: 3
}
```

I'm making this change to be able to find the exact line where the syntax error occurred. It still needs some more work for the `expected escape character " but reached the end of file` error because currently it always reports the last line of the input and doesn't say on which line the opening escape character appeared.

But I'm personally more interested in the case where escaped values aren't supposed to spill on subsequent lines. In my use case I'm processing lots of CSV files coming from hundreds of different users that were put together with ad-hoc encoding. In those files, each line of text represents a single object but oftentimes escape characters are not properly quoted in them or a single line may contain an invalid code point.

So I'm planning to add the ability to skip errored lines and continue parsing the remaining ones, accumulating errors along the way and returning them all after the whole input stream has been processed. I need to preserve the number of each faulty line to make it easier to find in the original file and report it to the author.

Reply:
Oh, you mean the newline character. Got it. I'll add the necessary tests and wil update this PR.

Reply:
I need a bit more help from you to validate my understanding.

In the current implementation, it doesn't matter much which value the `newlines` option has in `parse_enumerable` and `parse_stream`. Both functions treat every item of the enumerable as a separate row. This means that the behaviours of those two and `parse_string` are vastly different:

```
iex(1)> NimbleCSV.define(M, escape: "_", newlines: ["|"])                                  
{:module, M,
 <<70, 79, 82, 49, 0, 0, 40, 208, 66, 69, 65, 77, 65, 116, 85, 56, 0, 0, 2, 251,
   0, 0, 0, 69, 8, 69, 108, 105, 120, 105, 114, 46, 77, 8, 95, 95, 105, 110,
   102, 111, 95, 95, 7, 99, 111, 109, 112, ...>>, :ok}

iex(2)> M.parse_enumerable(["1,2|one,", "_two |,| four_", "|uno,dos"], skip_headers: false)
[["1", "2|one", ""], ["two |,| four"], ["|uno", "dos"]]

iex(3)> M.parse_enumerable(["1,2|one,_two |,| four_|uno,dos"], skip_headers: false)        
** (NimbleCSV.ParseError) unexpected escape character _ in "two |,| four_|uno,dos"
    lib/nimble_csv.ex:420: M.escape/6
    lib/nimble_csv.ex:290: anonymous fn/4 in M.parse_enumerable/2
    (elixir) lib/enum.ex:1075: anonymous fn/3 in Enum.flat_map_reduce/3
    (elixir) lib/enum.ex:3317: Enumerable.List.reduce/3
    (elixir) lib/enum.ex:1074: Enum.flat_map_reduce/3
    lib/nimble_csv.ex:229: M.parse_enumerable/2

iex(3)> M.parse_string("1,2|one,_two |,| four_|uno,dos", skip_headers: false)              
[["1", "2"], ["one", "two |,| four"], ["uno", "dos"]]
```

It also means that not just any stream can be passed to `parse_stream`. For example, `File.stream!("data.csv", [], 4096)` won't work. This is not explained in the docs. I'm not sure how your comments about new line characters in `@escape` apply here either.

I'm mostly interested in parsing streams because I have to deal with big files that should not be loaded into memory all at once. In the interest of efficiency I might want to tweak the chunk size in `File.stream!` but that's not compatible with NimbleCSV's `parse_stream`. Is this the intended behaviour of the library?

Reply:
> Is this the intended behaviour of the library?

Good point. The library started with binaries and streams were added later, so I think the line driven approach was by accident. In any case, it is what it is today, and we can assume newlines for streams, but we can't strings. Does that make sense?

Reply:
Ping. :)

Reply:
Sorry, I got distracted by other things at work. I'm going to come back to working on this issue either this or next week.

Reply:
Feel free to close this PR, I can always open another one when I have something new to show.